File: frontend/src/lib/services/settingsService.ts
import { apiClient, publicApiClient } from '@/lib/api';

export type TermsType = 'TERMS_OF_SERVICE' | 'PRIVACY_POLICY' | 'SHIPPING_RETURN_POLICY';

export interface BusinessSettings {
  id: number | null;
  companyName: string | null;
  representativeName: string | null;
  businessRegistrationNumber: string | null;
  mailOrderBusinessNumber: string | null;
  address: string | null;
  customerServicePhone: string | null;
  customerServiceEmail: string | null;
  brandName: string | null;
  updatedBy: number | null;
  createdAt: string | null;
  updatedAt: string | null;
}

export interface BusinessSettingsUpdateRequest {
  companyName: string;
  representativeName: string;
  businessRegistrationNumber: string;
  mailOrderBusinessNumber: string;
  address: string;
  customerServicePhone: string;
  customerServiceEmail: string;
  brandName: string;
}

export interface TermsVersion {
  id: number;
  type: TermsType;
  title: string;
  content: string;
  version: string;
  effectiveFrom: string;
  active: boolean;
  createdBy: number | null;
  createdAt: string;
}

export interface TermsVersionCreateRequest {
  type: TermsType;
  title: string;
  content: string;
  version?: string;
  effectiveFrom?: string;
}

export const TERMS_TYPE_LABEL: Record<TermsType, string> = {
  TERMS_OF_SERVICE: '\uC774\uC6A9\uC57D\uAD00',
  PRIVACY_POLICY: '\uAC1C\uC778\uC815\uBCF4\uCC98\uB9AC\uBC29\uCE68',
  SHIPPING_RETURN_POLICY: '\uBC30\uC1A1/\uBC18\uD488 \uC815\uCC45',
};

export const settingsService = {
  getCompanySettings: () => apiClient<BusinessSettings>('/admin/settings/company'),

  updateCompanySettings: (data: BusinessSettingsUpdateRequest) =>
    apiClient<BusinessSettings>('/admin/settings/company', {
      method: 'PUT',
      body: JSON.stringify(data),
    }),

  getTerms: () => apiClient<TermsVersion[]>('/admin/settings/terms'),

  createTerms: (data: TermsVersionCreateRequest) =>
    apiClient<TermsVersion>('/admin/settings/terms', {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  getLatestTerms: (type: TermsType) =>
    apiClient<TermsVersion>(`/admin/settings/terms/${type}/latest`),

  getTermsVersions: (type: TermsType) =>
    apiClient<TermsVersion[]>(`/admin/settings/terms/${type}/versions`),

  getTermsVersion: (type: TermsType, versionId: number) =>
    apiClient<TermsVersion>(`/admin/settings/terms/${type}/versions/${versionId}`),

  getPublicCompanySettings: () =>
    publicApiClient<Omit<BusinessSettings, 'id' | 'updatedBy' | 'createdAt' | 'updatedAt'>>('/settings/company/public'),

  getPublicLatestTerms: (type: TermsType) =>
    publicApiClient<Omit<TermsVersion, 'id' | 'active' | 'createdBy' | 'createdAt'>>(`/settings/terms/${type}/latest`),
};

File: backend/src/main/java/com/commerceops/erp/domain/settings/service/SettingsService.java
package com.commerceops.erp.domain.settings.service;

import com.commerceops.erp.domain.audit.enums.AuditActionType;
import com.commerceops.erp.domain.audit.service.AuditLogService;
import com.commerceops.erp.domain.settings.dto.BusinessSettingsResponse;
import com.commerceops.erp.domain.settings.dto.BusinessSettingsUpdateRequest;
import com.commerceops.erp.domain.settings.dto.PublicBusinessSettingsResponse;
import com.commerceops.erp.domain.settings.dto.PublicTermsVersionResponse;
import com.commerceops.erp.domain.settings.dto.TermsVersionCreateRequest;
import com.commerceops.erp.domain.settings.dto.TermsVersionResponse;
import com.commerceops.erp.domain.settings.entity.BusinessSettings;
import com.commerceops.erp.domain.settings.entity.TermsVersion;
import com.commerceops.erp.domain.settings.enums.TermsType;
import com.commerceops.erp.domain.settings.repository.BusinessSettingsRepository;
import com.commerceops.erp.domain.settings.repository.TermsVersionRepository;
import com.commerceops.erp.domain.user.entity.User;
import com.commerceops.erp.global.exception.BusinessException;
import com.commerceops.erp.global.exception.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class SettingsService {

    private final BusinessSettingsRepository businessSettingsRepository;
    private final TermsVersionRepository termsVersionRepository;
    private final AuditLogService auditLogService;

    public BusinessSettingsResponse getCompanySettings() {
        return businessSettingsRepository.findFirstByOrderByIdAsc()
                .map(BusinessSettingsResponse::from)
                .orElseGet(BusinessSettingsResponse::empty);
    }

    public PublicBusinessSettingsResponse getPublicCompanySettings() {
        return businessSettingsRepository.findFirstByOrderByIdAsc()
                .map(PublicBusinessSettingsResponse::from)
                .orElseGet(PublicBusinessSettingsResponse::empty);
    }

    @Transactional
    public BusinessSettingsResponse updateCompanySettings(BusinessSettingsUpdateRequest request, User actor) {
        BusinessSettings settings = businessSettingsRepository.findFirstByOrderByIdAsc()
                .orElseGet(() -> businessSettingsRepository.save(BusinessSettings.builder().updatedBy(actor).build()));
        BusinessSettingsResponse before = BusinessSettingsResponse.from(settings);
        settings.update(
                clean(request.companyName()),
                clean(request.representativeName()),
                clean(request.businessRegistrationNumber()),
                clean(request.mailOrderBusinessNumber()),
                clean(request.address()),
                clean(request.customerServicePhone()),
                clean(request.customerServiceEmail()),
                clean(request.brandName()),
                actor
        );
        BusinessSettingsResponse response = BusinessSettingsResponse.from(settings);
        auditLogService.record(
                actor,
                AuditActionType.BUSINESS_SETTINGS_UPDATED,
                "BUSINESS_SETTINGS",
                settings.getId(),
                null,
                "updated",
                "사업자 설정을 수정했습니다.",
                toCompanyAuditJson(before),
                toCompanyAuditJson(response),
                "{\"changedFields\":\"business_settings\"}"
        );
        return response;
    }

    public List<TermsVersionResponse> getTermsVersions() {
        return termsVersionRepository.findAllByOrderByTypeAscCreatedAtDesc().stream()
                .map(TermsVersionResponse::from)
                .toList();
    }

    public List<TermsVersionResponse> getTermsVersions(TermsType type) {
        return termsVersionRepository.findByTypeOrderByCreatedAtDesc(type).stream()
                .map(TermsVersionResponse::from)
                .toList();
    }

    public TermsVersionResponse getTermsVersion(TermsType type, Long versionId) {
        return termsVersionRepository.findByIdAndType(versionId, type)
                .map(TermsVersionResponse::from)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND));
    }

    public TermsVersionResponse getLatestTerms(TermsType type) {
        return TermsVersionResponse.from(findLatestTerms(type));
    }

    public PublicTermsVersionResponse getPublicLatestTerms(TermsType type) {
        return PublicTermsVersionResponse.from(findLatestTerms(type));
    }

    private TermsVersion findLatestTerms(TermsType type) {
        return termsVersionRepository
                .findFirstByTypeAndActiveTrueAndEffectiveFromLessThanEqualOrderByEffectiveFromDescCreatedAtDesc(type, LocalDateTime.now())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND));
    }

    @Transactional
    public TermsVersionResponse createTermsVersion(TermsVersionCreateRequest request, User actor) {
        TermsType type = request.type();
        termsVersionRepository.findByTypeAndActiveTrue(type).forEach(TermsVersion::deactivate);
        String version = clean(request.version());
        if (version == null) {
            version = "v" + (termsVersionRepository.countByType(type) + 1);
        }
        LocalDateTime effectiveFrom = request.effectiveFrom() != null ? request.effectiveFrom() : LocalDateTime.now();
        TermsVersion saved = termsVersionRepository.save(TermsVersion.builder()
                .type(type)
                .title(clean(request.title()))
                .content(request.content().trim())
                .version(version)
                .effectiveFrom(effectiveFrom)
                .active(true)
                .createdBy(actor)
                .build());
        auditLogService.record(
                actor,
                type == TermsType.TERMS_OF_SERVICE ? AuditActionType.TERMS_VERSION_CREATED : AuditActionType.POLICY_VERSION_CREATED,
                "TERMS_VERSION",
                saved.getId(),
                null,
                saved.getVersion(),
                type + " 새 버전을 생성했습니다.",
                null,
                null,
                "{\"type\":\"" + type + "\",\"version\":\"" + escapeJson(saved.getVersion()) + "\",\"title\":\"" + escapeJson(saved.getTitle()) + "\",\"contentLength\":" + saved.getContent().length() + "}"
        );
        return TermsVersionResponse.from(saved);
    }

    private String clean(String value) {
        if (value == null || value.isBlank()) {
            return null;
        }
        return value.trim();
    }

    private String toCompanyAuditJson(BusinessSettingsResponse response) {
        if (response == null || response.id() == null) {
            return "{}";
        }
        return "{"
                + "\"companyName\":\"" + escapeJson(response.companyName()) + "\","
                + "\"representativeName\":\"" + escapeJson(response.representativeName()) + "\","
                + "\"businessRegistrationNumber\":\"" + escapeJson(response.businessRegistrationNumber()) + "\","
                + "\"mailOrderBusinessNumber\":\"" + escapeJson(response.mailOrderBusinessNumber()) + "\","
                + "\"phone\":\"" + escapeJson(response.customerServicePhone()) + "\","
                + "\"email\":\"" + escapeJson(response.customerServiceEmail()) + "\""
                + "}";
    }

    private String escapeJson(String value) {
        if (value == null) {
            return "";
        }
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
